### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by boris-marinov**
Done except the functions for converting it to `Either` and (I think also good to have) `Maybe`.

Suggestion: instead of having `fromEither` function, isn't it better if we have `toValidation` method on the Either datatype (and vise-versa). This way you woudn't have to include a module in order to do the conversion, and also seems to me more correct in therms of design.

Looking forward for input on that. 
Also what did you have in mind for the Conversions module in https://github.com/origamitower/folktale/issues/5?

### CODE REVIEW

**Comment by boris-marinov** [comment]
Oh, I have to do the `concat` function as well.

**Comment by robotlolita** [comment]
Ah, so the Conversions module was there to avoid this problem of "do we have an Either.fromValidation, or do we have a Validation.toEither?", it also avoids problems with circular dependencies.

Otherwise we'd ideally have both `Either.fromValidation` and `Validation.toEither`.

**Comment by boris-marinov** [comment]
OK, here is a `concat` function (not sure how to handle concating several `Success` vals but I assume that the emphasis here is on accumulating errors).

I will work on the conversions in a separate branch.

**Comment by robotlolita** [comment]
Looks reasonable. I think we could also add `.empty()` as `Failure([])` and make it a monoid over errors.

**Comment by boris-marinov** [comment]
The funny thing is that an empty `Success` instance can serve as a `mempty` for a `Validation` monoid definition:
```
Failure(a).concat(Success()).equals(Failure(a))
```
Just an idea.

**Comment by boris-marinov** [comment]
The funny thing is that an empty `Success` instance can serve as a `mempty` for a `Validation` monoid definition:
```
Failure(a).concat(Success()).equals(Failure(a))
```
Just an idea.

**Comment by robotlolita** [comment]
Ah, that's interesting. We could make it `Success(x => x)`, so it's also an identity for the Applicative instance. Unless there's some law it breaks that I'm not aware of.

**Comment by vendethiel** [comment]
I think that's how it makes sense. If you have `mempty` be a `Failure([])`, it means `Success().concat(Validation.mempty())` would be `Failure`, right?

**Comment by boris-marinov** [comment]
@vendethiel , I thing robotlolita was suggesting  that `Failure([])` can serve as `mempty`, just for `Failure` not for  `Validation`.

@robotlolita, aren't the applicative laws concerned with just `ap` and `of`? 

Regarding the monoid laws note that if  `Validation.empty = () => Success(Something)` as I suggested then they get broken:
```
Success(a).concat(Validation.empty()).equals(Success(a)) // false
```
Actually we can fix this by extending the `concat` function of Success, so instead of
```
Success.concat = function(a) {
  return a
}
```
make it into
```
Success.concat = function(a) {
  a.equals(Validation.empty()) ? this : a
}
```
Conceptually, I think all `Success` values are `mempty`, but the interface does not work that way. :)

File: src/data/validation/core.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+Success.prototype[fl.ap]  = function(aValidation) {
+  assertValidation('Success#ap', aValidation);
+  return Failure.hasInstance(aValidation) ? aValidation
+  :      /* otherwise */                    aValidation.map(this.value);
+};
+
+Failure.prototype[fl.ap]  = function(aValidation) {
+  assertValidation('Failure#ap', aValidation);
+  return Failure.hasInstance(aValidation) ? Failure(this.value.concat(aValidation.value))
+  :      /* otherwise */                    aValidation.map(this.value);
```
Comment (robotlolita) [inline src/data/validation/core.js]: This should just return `this` if `aValidation` is a Success.

File: src/data/validation/core.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+
+// -- Folds and extended transformations--------------------------------
+
+Validation.fold = function(f, g) {
+  return this.cata({
+    Success: ({ value }) => f(value),
+    Failure: ({ value }) => g(value)
+  });
```
Comment (robotlolita) [inline src/data/validation/core.js]: `Failure` should apply `f` here, and `Success` apply `g`, so we get the `Left = Failure, Right = Success` equivalence

File: src/data/validation/core.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+Validation.merge = function() {
+  return this.value;
+};
+
+Validation.swap = function() {
+  return this.fold(Failure, Success);
+};
```
Comment (robotlolita) [inline src/data/validation/core.js]: Same consideration as above.

File: src/data/validation/core.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+Validation.bimap = function(f, g) {
+  return this.cata({
+    Success: ({ value }) => Success(f(value)),
+    Failure: ({ value }) => Failure(g(value))
+  });
+};
```
Comment (robotlolita) [inline src/data/validation/core.js]: Same consideration as above

**Comment by boris-marinov** [comment]
Fixed the issues that you found.
Fixed the isssue with the monoid laws breaking.

**Comment by robotlolita** [comment]
Hm, this is a confusing Monoid instance to me:

```
Success(1).concat(Validation.empty()) = Success(1)
Success(1).concat(Success(x => x)) ≃ Success(x => x)
Success(x => x) ≃ Validation.empty()
```

Granted function equivalence is a tricky thing. Maybe leave this out of the default Validation for now and revisit it later? I've got some ideas for multiple instances that I still have to write about.

**Comment by boris-marinov** [comment]
The aforementioned problem with conforming to monoid laws, is already fixed.

Do you still think that we should leave out the `empty`.

**Comment by robotlolita** [comment]
Ah, sorry, I could've been more clear. What I meant is that the monoid instance is valid, but it's confusing to me because it relies on identity rather than structural equality, so I'd prefer to leave it out of the default.

I'd be okay with providing a separate instance that people can import/opt-in for explicitly, but we don't have a good way of doing that yet.

**Comment by boris-marinov** [comment]
I agree.

**Comment by robotlolita** [comment]
Thanks :)

**Pull request merged**